### ISSUE TITLE
An existing contact could not be found using the specified contact details. The contact name field is required to create a new contact.

### ISSUE THREAD

**Issue by geoffpkr** [eileenmcnaughton/nz.co.fuzion.civixero#98]
Invoice error - how to fix??

"An existing contact could not be found using the specified contact details. The contact name field is required to create a new contact."

**Comment by geoffpkr** [comment]
I don't think the entry in Xero Community https://community.xero.com/developer/discussion/102396526 regarding wptransients is relevant as our CiviCRM site is hosted in Joomla/

**Comment by eileenmcnaughton** [comment]
It's definitely a civi issue - the message suggests you are pulling in a contact from Xero and it doesn't have enough detail to match it? Or pushing to Xero - I'm guessing name/email details might be missing?

**Comment by petednz** [comment]
does the civi contact connected to this contribution show a connect to Xero on their Summary screen or does it show 'queue sync to Xero' - if it does, and you click on it, does it change to 'Contact is queued for sync with Xero". if it does and you then run the Scheduled Job for CiviXero Contact Push Job what result do you get?
if nothing has changed, what do you see at /civicrm/xero/errorlog?

Note - i have found one example of this error in one client site. It is odd. The contributions are synched but the Contact is not ie there is a Contact in Xero with the correct civi contact ID, but the contact in civi does not show it is linked, even though their contributions do.

The values in civicrm_accounts_invoice.accounts_data correctly shows the civi contact id and the xero contact id

I can't explain 
a/ why there is now row in civicrm_account_contact that matches the above
b/ why setting the civi contact to queue does not result in a record being created in civicrm_account_contact 

ah unless it is due to "Oath rate exceeded" - which iirc correctly can happen if there are too many contacts in civi which keep 'retrying' to connect to xero (but doesn't explain how come we do not have a row in civicrm_account_contact.

Hmm. checking log_civicrm_account_contact for same contact id - now i see a LOT of rows of info - none of which are marked as Delete so again not clear what might have precipitated it being removed - let me quiz my colleague about this and come back to you next week (he is away this week)

**Comment by geoffpkr** [comment]
Three of the four instances show "Contact is queued for sync with Xero"
In one of the four instances there were 2 duplicate contacts in Civi that I have now fixed - the correct remaining contact is correctly linked to Xero.
I have run the Scheduled Job for CiviXero Contact Push Job, and nothing has changed.
I also ran the Scheduled Job for CiviXero Invoice Push Job, and nothing has changed 

I will send the contents of /civicrm/xero/errorlog separately

**Comment by petednz** [comment]
sure. as i said
> let me quiz my colleague about this and come back to you next week
since i have an example that may match your situation (and which is clearly not correct)

**Comment by geoffpkr** [comment]
"I will send the contents of /civicrm/xero/errorlog separately"

Nothing in /civicrm/xero/errorlog

**Comment by petednz** [comment]
Looking at the example I found, my colleague said

> I think this is because connector_id field is set to NULL so I have updated it  via sql query to store 0 for the connector_id field in the civicrm_account_contact table.

> I think this is the field value from which the extension decides which Xero account to connect to if there are multiple Xero accounts involved. Since we have only 1 xero account here, it should be set to 0 for all rows.

However, I then checked on another site and found examples of contacts who had multiple rows in the _account_contact table, all with NULL in the connector_id field. so maybe there is a bigger issue here.

**Comment by geoffpkr** [comment]
The problem persists after multiple retries so presumably is not server timeout

**Comment by petednz** [comment]
Does "contact ID appended eg John Smith - 2345" have NULL or 0 in the connector_id field?
If it has NULL and you update it to 0 then does anything get resolved for that contact?

**Comment by geoffpkr** [comment]
I tried each of the 4 contacts - all had NULL in the connector_id field.
I changed this to 0 and got the following error
"#1062 - Duplicate entry '6722-0-xero' for key 'UI_contact_id_plugin'"
for each of the 4 contacts
Have reverted to NULL for now

**Comment by petednz** [comment]
Did each contact only have one row in that table, or did they all have multiple?

**Comment by geoffpkr** [comment]
each of the 4 contacts had only one entry in civicrm_account_contact table.

We have ~6000 contacts and many have NULL in the connector_id field (and appear to synch normally), but some have 0 in that field.

In the error message, #1062 - Duplicate entry '6722-0-xero' for key 'UI_contact_id_plugin',  6772 is the Contact ID for that particular example.

**Comment by geoffpkr** [comment]
I mean 6722 is the Contact ID

**Comment by geoffpkr** [comment]
I have searched in civicrm_account_contact table for contact_id 6722 and found a meaningless duplicate !!!! for that contact_id which I have deleted, and similar for the other 3 contact_id's.  These duplicates had no information except that contact_id
Will see if sync occurs now.

**Comment by geoffpkr** [comment]
No does not sync - creates another "meaningless" entry in civicrm_account_contact but does not sync with the real contact

**Comment by geoffpkr** [comment]
Does "contact ID appended eg John Smith - 2345" have NULL or 0 in the connector_id field?
If it has NULL and you update it to 0 then does anything get resolved for that contact?

For contact 6722, which has two entries in the table, one containing relevant contact info, the connector_id field  for the main entry had NULL, I updated to 0 but no change with syncing that contact or invoice

**Comment by geoffpkr** [comment]
Still having difficulty with this problem
I have 4 existing contacts and two new contacts that will not synch to Xero
I have noticed that the Contact Card in Xero for working contacts has a Contact Code field in the Card, whereas those  who won't sync have no Contact Code field

Working contact
![Screen Shot 2022-10-01 at 5 04 32 pm](https://user-images.githubusercontent.com/46299874/193397891-ce74bf1a-aa6c-4ebf-a801-90ea6970dc17.png)

Non-synching contact
![Screen Shot 2022-10-01 at 5 09 53 pm](https://user-images.githubusercontent.com/46299874/193397906-a1156c8b-8485-493b-96be-ffff3e2c5f88.png)

Not certain if this might help

**Comment by geoffpkr** [comment]
The problematic contacts persistently show "Contact is queued for sync with Xero" in CiviCRM but the sync does not occur - - despite XeroAuthorize saying 
CiviCRM can connect to Xero. You do not need to authorize again at this point.
and other tasks flowing back and forth to Xero
Seems that the contact sync is blocked at some point. Is is possible to clear the queue and start again?

**Comment by geoffpkr** [comment]
Repeated error message in /media/civicrm/ConfigAndLog

Oct 08 21:25:02  [warning] hook_civicrm_accountsync_plugins should be updated to pass an array of parameter names to CRM_Utils_Hook::invoke(). Caller: ::_accountsync_get_enabled_plugins
Array
(
    [civi.tag] => deprecated

**Comment by geoffpkr** [comment]
Would really like to get this resolved - happy to make a donation

**Comment by geoffpkr** [comment]
I have kept trying and because the Invoice Push Job error message refers to a Contact Push error, I made a few minor changes in the Contact ID's in Xero and the Contact Push job then ran flawlessly.

I hoped that the Invoice Push job would then work properly but on the next run, the same 15 errors occurred as above (btw more than 300 had worked properly).

       • A validation exception occurred
       • An existing contact could not be found using the specified contact details. The contact name field is required to create a new contact.

It seemed to me that the Invoices had been flagged in some way so that they would not transfer.

I then cleared these errors in CiviCRM and created the relevant invoices manually in Xero with the Invoice Numbers from CiviCRM and sent them off. 

This led to 15 errors in the next run of the Invoice Pull Job as below when Xero sent those invoices back to CiviCRM for synchronization, but this is not a big problem because those invoices started in CiviCRM.

I think there must be a simple solution to this, but I wish I knew what it was.

----------------------------------------------------------------

plugin=xero
start_date=yesterday

Parameters parsed (and passed to API method): 
a:3:{s:7:"version";i:3;s:6:"plugin";s:4:"xero";s:10:"start_date";s:9:"yesterday";}

Full message: 
Finished execution of CiviXero Invoice Pull Job with result: Failure, Error message: Not all records were savedArray
(
    [0] => Failed to store INV_2765 (a2658b27-21dd-4fae-88b5-980b2f47354e ) with error contribution_id is not a valid integerInvoice Pull failed
    [1] => Failed to store INV_2766 (e53417f2-4eb5-4fed-91be-17ade6f03bbc ) with error contribution_id is not a valid integerInvoice Pull failed
    [2] => Failed to store INV_2844 (2ab67738-fd4c-4e0d-b59e-6d0d9737a2dc ) with error contribution_id is not a valid integerInvoice Pull failed
    [3] => Failed to store INV_2767 (b9915607-2e21-4a38-be23-4db5010def73 ) with error contribution_id is not a valid integerInvoice Pull failed
    [4] => Failed to store INV_2768 (5982ecfc-073b-46c5-b156-93ba5aaf7c72 ) with error contribution_id is not a valid integerInvoice Pull failed
    [5] => Failed to store INV_2774 (f20547f5-37c1-4600-88fe-85cf3efcd355 ) with error contribution_id is not a valid integerInvoice Pull failed
    [6] => Failed to store INV_2853 (b9026472-6725-45f9-a79f-df4e71de134a ) with error contribution_id is not a valid integerInvoice Pull failed
    [7] => Failed to store INV_2892 (0ff61958-70de-47fa-8228-7548aecd4055 ) with error contribution_id is not a valid integerInvoice Pull failed
    [8] => Failed to store INV_2911 (f3972310-c9d4-4375-a595-5cf80ecb4cd3 ) with error contribution_id is not a valid integerInvoice Pull failed
    [9] => Failed to store INV_2942 (82dbac48-6598-46bb-915f-cf482bc1a2fb ) with error contribution_id is not a valid integerInvoice Pull failed
    [10] => Failed to store INV_2955 (1041b98c-e173-4e2d-8095-89c88f428d9c ) with error contribution_id is not a valid integerInvoice Pull failed
    [11] => Failed to store INV_2959 (f023fae7-2d55-47e6-8544-bd23278d7be6 ) with error contribution_id is not a valid integerInvoice Pull failed
    [12] => Failed to store INV_3025 (c73ff985-d535-4f99-b2a3-65c877e2664a ) with error contribution_id is not a valid integerInvoice Pull failed
    [13] => Failed to store INV_3039 (8d3b5e0d-43a9-432c-843f-3b098246233f ) with error contribution_id is not a valid integerInvoice Pull failed
    [14] => Failed to store INV_3060 (df046d8d-511d-4251-93c3-886f292d8b56 ) with error contribution_id is not a valid integerInvoice Pull failed
    [15] => Failed to store INV_3065 (efe16448-cda9-4067-b69b-6fb7cc22b674 ) with error contribution_id is not a valid integerInvoice Pull failed
)

**Comment by eileenmcnaughton** [comment]
Is this on master (master is a little bit in flux at the moment)

**Comment by geoffpkr** [comment]
5.56.1
Joomla 3.10

**Comment by eileenmcnaughton** [comment]
ok- but for civixero?

**Comment by geoffpkr** [comment]
We are using 2.02 - hope that is what you are looking for

**Comment by geoffpkr** [comment]
Do I need all patches installed to be “Master”

**Comment by eileenmcnaughton** [comment]
Yeah - the version from https://github.com/eileenmcnaughton/nz.co.fuzion.civixero/archive/refs/heads/master.zip

**Comment by geoffpkr** [comment]
Will install Sent from my iPhoneOn 13 Jan 2023, at 8:37 am, Eileen McNaughton ***@***.***> wrote:﻿
Yeah - the version from https://github.com/eileenmcnaughton/nz.co.fuzion.civixero/archive/refs/heads/master.zip

**Comment by geoffpkr** [comment]
should I disable or delete 2.02 first

**Comment by eileenmcnaughton** [comment]
@geoffpkr remove the code base first - 

@mattwire perhaps I should tag a release of this & the other - we seem to be in a stable state right now - I have the latest master runing on prod

**Comment by mattwire** [comment]
> @geoffpkr remove the code base first -
> 
> @mattwire perhaps I should tag a release of this & the other - we seem to be in a stable state right now - I have the latest master runing on prod

I'm happy with that - but the issue with `isApiRateLimitExceeded()` needs fixing first. In your version you moved the exception out of the function (I'm fine with deprecating/removing the parameter but it would be nice to have a single function/line to call at the beginning of each action which can trigger the exception directly).
The problem is you can't call the XeroThrottle exception again while the rate limit is exceeded because it will reset the rate limit...

**Comment by eileenmcnaughton** [comment]
@mattwire OK - OK - so I found that the setting of the setting was hidden in the `__construct` function - there are 2 places that throw that Exception, one of which does it with the intent to set the setting (and to log that it got throttled) - so I moved that handling back out of the exception construct & into the validate - which should mean both places can use that exception for their slightly different purposes

**Comment by geoffpkr** [comment]
We have done the upgrade to master (showing as V2.1 in the Extensions list) 
"Installed master, but without commit 190a09a9 because it caused the following error on CiviCRM's extension admin page (civicrm/admin/extensions):
Call to undefined function _civixero_civix_civicrm_upgrade()"

We also upgraded the Account Sync extension to the current master

This is now seen in the CiviCRM system status after updating the Account Sync extension:
AccountSync: Database issues
"There are 838 records in the `civicrm_account_contact` table which have a NULL connector_id.
These need updating manually to 0 or the connector ID if using the connectors extension"

I have also noted that under Administer menu, the Xero Authorise and Xero Settings options are visible, but the Xero Invoice Errors, Contact errors and Synchronise contacts menu options are not visible now.

**Comment by geoffpkr** [comment]
Clean-up of the 838 rows in the `civicrm_account_contact` table which have a NULL connector_id.

First, deleted rows where connector_id and accounts_contact_id were both null. These were empty rows that did not map anything. DELETE FROM `civicrm_account_contact` where accounts_contact_id is null

Second, AT cleaned up table by manually deleting rows with null connector_id where another row exists with same contact ID. 

The remainder (80 rows) all appear to be members.

These appear to just have null connector_id but no other problem (at least not obvious problems). Corrected these by setting to 0.

All of the 838 missing connector_id have now been resolved. That error no longer appears. There are now a large number of contacts that are marked for syncing i.e. accounts_needs_update = 1.

CiviXero Contact Push Job
Finished execution of CiviXero Contact Push Job with result: Failure, Error message: invalid_grant

This may be a Xero error as described in https://developer.xero.com/community-forum-archive/discussion/127920775

**Comment by geoffpkr** [comment]
Xero Help Desk replied
Hi Geoffrey

Thank you for sending your client id.

I can see that refresh token 7CJ0****UWsA is being used many times to refresh the access token, 479 times so far.

This token was swapped for token N4qV****37RM with timestamp 01/18/2023 08:46:51.000 +0000 and so the 7CJ0****UWsA was marked as expired.

To resolve this issue, you would need to go through the steps to connect the app to Xero again as this will generate a new access and refresh token.

Please note, if the refresh fails, going forward it would be better to change the code so that the app doesn't keep trying. Implementing token logging would be helpful so that the app could then move to use the most recent token.

Kind regards

**Comment by geoffpkr** [comment]
I have deleted and recreated the Civi CRM OAuth 2.0 app in Xero with new Client ID and secret and entered in CiviCRM

Inserted the redirect URL taken from the Xero Authorise page in CiviCRM which is 
https://www.prr.org.au/membership/administrator/?option=com_civicrm&task=civicrm/xero/authorize&connector_id=0

I found that I got an Error 500 "invalid Redirect URL" when trying to Authorise Xero until I deleted "&connector_id=0" from the end of the Redirect URL, and then the Xero Authorise function worked properly with the new Client ID = DDFC4BD465194DDFA0616082E34D67BA

On running the sync jobs, I am still getting the error message
Finished execution of CiviXero Contact Push Job with result: Failure, Error message: invalid_grant
AT says
"I have confirmed that CiviCRM is still attempting to use refresh token 7CJ0****UWsA, which Xero support says has expired, in nz.co.fuzion.civixero/CRM/Civixero/OAuth2/Xero.php renewToken().

Further notes:

$token->getRefreshToken() returns 7CJ0****UWsA

This line in the renewToken function:
$newToken = $this->provider->getAccessToken('refresh_token', [
      'refresh_token' => $token->getRefreshToken()
    ]);

is actually crashing silently. I can't get it to log any error messages and no code after $newToken = ... is executed (when using the expired refresh token - so this is not the cause of the problem but no error is being generated).

**Comment by geoffpkr** [comment]
Hi Geoffrey

Thank you for the timestamp.

If you receive a Invalid Grant Error, the most likely cause will be that the Refresh token you're using is either invalid, expired or has previously been used. I have checked our logs for the time you mention and can confirm that this is the reason for your error. The expired refresh token 7CJ0****UWsA was used in a refresh call.

The easiest way to resolve this error, is to simply re-authorise the App with the Xero Organisation.

We'd suggest that the developer team for the app add a token logging mechanism to log each of the Tokens the app is using and to confirm if the app is using the correct token each time. 

For more information on Invalid Grants and troubleshooting these errors, please see the link below. 

Xero Developer Centre: [Troubleshooting](https://developer.xero.com/documentation/oauth2/troubleshooting)

Kind regards

Sally

**Comment by eileenmcnaughton** [comment]
OK - this is getting super confusing with versions. I have

1) merged a patch by @mattwire that seeks to help with token management
2) fixed the dangling upgrade functions (those calls to the non existent upgrade functions
3) tagged a new release

It's still possible we need another release fairly soon as I can't recall where we got to on the throttle

**Comment by geoffpkr** [comment]
That is a great help - It looks like the OAuth2.0 account access is now working - the Access token is accepted - however, the Refresh token does not seem to be automatically presented to Xero so Authorisation needs to be re-done manually every 30 mins
also, it seems that the Auth status reported by 
https://.... administrator/?option=com_civicrm&task=civicrm/a/#/status
and 
https://.... administrator/?option=com_civicrm&task=civicrm/xero/authorize&connector_id=0

are not always in agreement despite refreshing the page.

**Comment by eileenmcnaughton** [comment]
@mattwire do you know where this is at? I tagged a release with what is merged to date to reduce confusion

**Comment by mattwire** [comment]
@eileenmcnaughton @geoffpkr Since my latest fixes for oauth tokens I've had no problems with `invalid_grant` and the scheduled jobs have worked every time.
However, I think there is an issue with the logic driving the system status message reporting that "Authorization needs to be re-done". When the scheduled job runs that message goes away but then comes back again. The problem I think is that the access token *is* only valid for 30 minutes but a new access token can be obtained from the refresh token for up to 60 days - if *that* fails you do need to re-authorize but I'm not sure how to detect that scenario.

**Comment by geoffpkr** [comment]
Yes I agree that is exactly what happens, except that the invalid_grant message appears transiently if a contact sync is attempted when the authorization is not working.

**Comment by geoffpkr** [comment]
We have installed the latest commits
Unfortunately the Access token still only lasts for 30 mins, and the refresh token does not seem to be working.

Also, have noted CiviXero Menus items are now in 
Administer/CiviContribute/Accounts Integration

I can authorize through https:// ....... /administrator/?option=com_civicrm&task=civicrm/xero/authorize
but can't find XeroAuthorize on the menus

**Comment by mattwire** [comment]
@geoffpkr For the menus you may need to clear caches a couple of times! Or even try rebuilding menu https://docs.civicrm.org/sysadmin/en/latest/troubleshooting/#rebuild-civicrm-menus

**Comment by geoffpkr** [comment]
Thanks  - in my case was hidden below bottom of screen (!!!) below Xero Settings - visible in Full Screen Mode

**Comment by geoffpkr** [comment]
I keep trying and have deleted the Civi OAuth 2.0 app in Xero developer and reauthorized with a new Client ID and Client Secret 
Authorization is straightforward but the token always lasts for 30 mins and the refresh token does not seem to work.  2.1.0 and recent commits - is there anything we can do about this ??

**Comment by mattwire** [comment]
I'm seeing issues on one site - have a look at https://github.com/eileenmcnaughton/nz.co.fuzion.civixero/pull/101/commits/c4acc25084ac7c72e03b922de4f2b79761f9cd12 - not sure if it fully fixes the problem yet

**Comment by mattwire** [comment]
Ok, see https://github.com/eileenmcnaughton/nz.co.fuzion.civixero/pull/126

**Comment by geoffpkr** [comment]
I have done the Edits from Matt on 5 files and the lines from Eileen in TokenStoreDefault.php
When I go to Accounts Integration/XeroAuthorize the link to Xerol does not seem to be triggered (white CiviCRM screen only) 
I can't find any errors in the edits I have made

**Comment by geoffpkr** [comment]
I was able to re-authorize after AT moved to current master (as at commit 04e7e08) because the Xero Authorize page was again visible, but the Authorization again worked only for 30 mins and it appeared the Refresh token was not working properly.  
I tried creating a new client secret in CiviCRM OAuth 2.0 in Xero Developer but I now get an error on the Xero Authorize page 
'Sorry, due to an error, we are unable to fulfill your request at the moment. You may want to contact your administrator or service provider with more details about what action you were performing when this occurred.
Required option not passed: "access_token" '

I removed and reinstalled the CiviXero extension (to V 2.1.1) but get the same error.

**Comment by eileenmcnaughton** [comment]
How often does your cron run? What I see is that when we run a cron within 30 minutes it pushes the refresh date out - presumably meaning that if your cron runs less often it might not work as well as if it runs more often (it runs all the time on the site I just checked)

**Comment by geoffpkr** [comment]
I believe every 5 mins but will confirm

**Comment by geoffpkr** [comment]
yes is 5 mins

**Comment by mattwire** [comment]
@geoffpkr I've not seen any issues since the latest updates on any sites. Are you using connectors or just civixero+accountsync?

**Comment by geoffpkr** [comment]
Hi Matt - the Redirect URL is https://xxxxxx/membership/administrator/?option=com_civicrm&task=civicrm/xero/authorize&connector_id=0
Hope that helps

**Comment by geoffpkr** [comment]
The problem occurred when I renewed the client secret because I thought that might help.  I can't now get to the page to enter the new Client ID or Client secret - might explain error

**Comment by geoffpkr** [comment]
I have upgraded to 2.1.2 but am still getting the below error when trying to authorize
Required option not passed: "access_token"

**Comment by geoffpkr** [comment]
I have upgraded to 2.1.3 but am still getting the below error when trying to authorize
Required option not passed: "access_token"

**Comment by geoffpkr** [comment]
I have deleted and re-created the CiviXero 0Auth 2 app
I still cannot access the Xero Authorization page

The message in CiviCRM has changed to say  
"Missing Xero App Details
Please configure a Client ID and Client Secret from your Xero app."
whereas it used to say "Xero Authorization failure 
Please reauthorize etc

I have the new User Id and secret, but I can't get to the Xero Authorize page in CiviCRM to enter the new UserID and Client secret.  Happy to make a contribution if can help.

**Comment by geoffpkr** [comment]
AT says "The error that appears when you open the CiviXero Authorise page (access_token) looks to me like a Xero OAuth error (its API is not being supplied the access token by CiviXero), but I do not know why.

**Comment by eileenmcnaughton** [comment]
@geoffpkr are you on CivXero 2.1.3 - I released that a few days ago & someone said if fixed their auth issue

**Comment by geoffpkr** [comment]
Yes we are on CiviXero 2.1.3.

AT says " Yesterday I deleted the CiviXero settings from the SQL database but it did not help so I believe there is a bug that stops the CiviXero authorise page from working.  I have put the settings back as they were however you are still not going to be able to authorise."

The Redirect URL is https://www.xxxxx.org.au/membership/administrator/?option=com_civicrm&task=civicrm/xero/authorize&connector_id=0  and I have confirmed this is an absolute URL as required by the OAuth 2.0 spec (section 3.1.2 of RFC 6749), and that we are not being blocked by a ModSecurity rule.  The Redirect URL worked previously.

**Comment by eileenmcnaughton** [comment]
@geoffpkr so when you choose to authorize through the UI what do you experience?

**Comment by eileenmcnaughton** [comment]
Oh I see "I have upgraded to 2.1.3 but am still getting the below error when trying to authorize
Required option not passed: "access_token"" - but you are an intermediary here?

**Comment by eileenmcnaughton** [comment]
I'm still a bit confused by " I can't now get to the page" - is that when you hit the url for authorize?

**Comment by geoffpkr** [comment]
When I go to the page https://www.xxxxx.org.au/membership/administrator/?option=com_civicrm&task=civicrm/xero/authorize&connector_id=0, CiviCRM returns an error page that states "Required option not passed: "access_token"
I cannot get to the screen where I can enter the User ID and Client Secret

**Comment by eileenmcnaughton** [comment]
@geoffpkr I just pushed up a patch that seems to help - in master branch

**Comment by geoffpkr** [comment]
Thanks - will try

**Comment by geoffpkr** [comment]
Great  - definitely progress here - with the patch installed I can get to the Xero Authorize page and enter the new Client ID and Client Secret.
The Access token expiry date is still showing as 9 April 2023.

When I press the Authorize with Xero button, I get the following error

Error code: 500
Error: unauthorized_client : Invalid redirect_uri

ID and Secret are correct

**Comment by eileenmcnaughton** [comment]
@geoffpkr I've seen that invalid redirect url - it occurs when the url configured in xero is different to what the code tries to redirect to - I think it might be because it is not now adding `&connector_id=0' to the url but you might have that configured in xero (if you get it working perhaps you could put up a docs update)

**Comment by geoffpkr** [comment]
I removed "&connector_id=0" from the Redirect URL in Xero app and can authorize !!!
Thanks so much 
I will watch for the refresh token in 30 mins

**Comment by eileenmcnaughton** [comment]
great!

**Comment by geoffpkr** [comment]
There was a hitch at the time of token refresh with a CiviCRM error 

Return value of CRM_Civixero_OAuth2_TokenStoreDefault::getConnectorID() must be of the type int, null returned

but this seemed to clear when I pressed Authorize again

Scheduled jobs now appear disabled

Data is synchronized via scheduled jobs. No data will be synchronized until you enable one or more jobs.
Push updated contacts to Xero: Disabled
Pull updated contacts from Xero: Disabled
Push updated invoices from Xero: Enabled
Pull updated invoices from Xero: Disabled
Complete Contributions in CiviCRM where completed in Accounts: Disabled

**Comment by geoffpkr** [comment]
The disabled jobs were duplicates and have been deleted

**Comment by eileenmcnaughton** [comment]
@geoffpkr hmm - that sounds like the error that was in 2.1.2 - I did cut 2.1.4 so maybe re-download to be sure you have it

**Comment by geoffpkr** [comment]
The current status is that the Xero Authorize page reports a valid Access Token with an expiry time in the future and  "CiviCRM can connect to Xero. You do not need to authorize again at this point."
All jobs enabled

When I first log in to CiviCRM, I get an incorrect message saying that I need to reauthorize with Xero, but this is not the case according to the XeroAuthorize page

**Comment by eileenmcnaughton** [comment]
@geoffpkr  is it this message?

![image](https://user-images.githubusercontent.com/336308/236337153-628bcda9-a47a-45a6-a571-49284537ef68.png)


or

![image](https://user-images.githubusercontent.com/336308/236337197-fae13484-fa59-4e5a-b98d-7a8a80c905a0.png)

**Comment by geoffpkr** [comment]
Hi Eileen
Please see attached - I think it is the critical but not sure

![IMG_1831](https://user-images.githubusercontent.com/46299874/236589180-e6454611-8d1a-40f9-a138-bf65378f5be0.jpg)

Top RHS Dialog Box

System Status: Critical

- Xero Authorization Required

Refresh token working beautifully

How do I update to 2.1.4, as the update availability is not advertised on the CiviCRM Extensions page?

**Comment by eileenmcnaughton** [comment]
I mistagged it - is should release 2.1.4 now

**Comment by geoffpkr** [comment]
Thank you - 2.1.4 is installed and working well.
Occasionally I am seeing the Xero Authorization Required notice on signing in but the refresh token has always worked when I go to the Xero Authorize page.

**Comment by eileenmcnaughton** [comment]
@geoffpkr yay - I'll close this & we can start a new thread with anything ongoing from here since we seem to have hit a stable point
